Recipe for NodeJS example (#8096)
microsoft/FluidFramework: docs/content/docs/recipes/node.md

---
title: Using Fluid with NodeJS
menuPosition: 3
author: sdeshpande3
aliases:
  - "/start/nodejs-tutorial/"
---

In this tutorial, you'll learn about using the Fluid Framework by building a simple application in NodeJS that enables connected clients to generate random numbers and display the result of any changes to the shared state.  You'll also learn how to connect the Fluid data layer in [Node](https://nodejs.org/).

To jump ahead into the finished demo, check out the [Node demo in our FluidExamples repo](https://github.com/microsoft/FluidExamples/tree/main/node-demo).

The following image shows the random number generated open in four terminals after every one second.

> ![Four terminals with the Random Number app open in them.](https://user-images.githubusercontent.com/46719950/139499122-76edb832-50ec-4c92-b6b7-06869e9bef9f.png)

Reviewer: the second terminal from this image has different numbers from the other 3 after like the first 4/5ths of them, and they're not just out of order either

Reviewer 2: thanks for noticing this. I have updated the image. Not sure why the second terminal was not updating the values properly.

   }

    async function createContainer() {
      // TODO 5: Create the container
    }

    async function loadContainer(id) {
      // TODO 6: Get the container from the Fluid service
    }

    async function start() {
        // TODO 7: Read container id from terminal and create/load container
    }

   start().catch(console.error());
    ```

1. Replace `TODO 1` with the following code. Note that there is only one object in the container: a SharedMap holding the random number. Note also that `sharedRandomNumber` is the ID of the `SharedMap` object and it must be unique within the container.

    ```js
      const client = new TinyliciousClient();
      const containerSchema = {
          initialObjects: { sharedRandomNumber: SharedMap }
      };
    ```

1. Replace `TODO 2` with the following code. Note that `containerId` is taken as an input, and if there is no `containerId` we create a new container instead.

    ```js
    let containerId = "";
    console.log("Type a Container ID or press Enter to continue: ");
    await readlineAsync().then( line => {
        console.log("You entered: " + line);
        containerId = line;
    });
    return containerId;
    ```

> 1. To ensure that both local and remote changes to the random number are reflected in the terminal, we will use the `newRandomNumber()` function to store the local value and `updateConsole()` function to ensure that the map is updated whenever any client changes the value. This helps in keeping the client terminals synchronized with the Fluid data.

Reviewer: ```suggestion
1. To ensure that both local and remote changes to the random number are reflected in the terminal, we will use the `newRandomNumber()` function to store the local value and `updateConsole()` function to ensure that the console is updated whenever any client changes the value. This helps in keeping the client terminals synchronized with the Fluid data.
```

   Replace `TODO 3` with the following code. It will set a timer to update the random number every one second.

   ```js
   const newRandomNumber = () => {
     map.set("random-Number-Key", Math.floor(Math.random() * 100) + 1);
   };
   setInterval(newRandomNumber, 1000);
   ```

>   Replace `TODO 4` with the following code. It will Llsten for updates and print changes to the random number.

Reviewer: ```suggestion
  Replace `TODO 4` with the following code. It will Listen for updates and print changes to the random number.
```

   ```js
   const updateConsole = () => {
     console.log("Value: ", map.get("random-Number-Key"));
   }
   updateConsole();
   map.on("valueChanged", updateConsole);
   ```

1. Replace `TODO 5` with the following code.

    ```js
    const { container } = await client.createContainer(containerSchema);
    container.initialObjects.map.set("random-Number-Key", 1);
    const id = await container.attach();
    console.log("Initializing Node Client----------", id);
    loadCli(container.initialObjects.map);
    return id;
    ```

1. Replace `TODO 6` with the following code.

   ```js
   const { container } = await client.getContainer(id, containerSchema);
   console.log("Loading Existing Node Client----------", id);
   loadCli(container.initialObjects.map);
   ```

> 1. Replace `TODO 7` with the following code. Note that, this code will first take the container id as the input. If a new Fluid container is created, press Enter or type `undefined`. A new container will be initialized and the container id will be printed in the terminal. You can copy the container id, launch a new terminal window, and type/paste the initial container id to have multiple collaborative NodeJS clients.

Reviewer: ```suggestion
1. Replace `TODO 7` with the following code. Note that this code will first take the container id as the input. To create a new Fluid container, press Enter or type `undefined`. A new container will be initialized and the container id will be printed in the terminal. You can copy the container id, launch a new terminal window, and type/paste the initial container id to have multiple collaborative NodeJS clients.
```
